smartfs.lua

```diff
@@ -326,31 +341,68 @@ function smartfs._makeState_(form, newplayer, params, is_inv, nodepos)
 				meta:set_string("smartfs_name", self.def.name)
 			end
 		end,
-		-- Receive fields and actions from formspec
-		_sfs_on_receive_fields_ = function(self, player, fields)
-			assert(self)
-			assert(player)
-
-			for key,val in pairs(fields) do
-				if self._ele[key] then
-					self._ele[key].data.value = val
+		-- process /apply received field value
+		_sfs_process_value_ = function(self, field, value)
+			local cur_namespace = self:getNamespace()
+			if cur_namespace == "" or cur_namespace == string.sub(field, 1, string.len(cur_namespace)) then -- Check current namespace
+				local rel_fieldname = string.sub(field, string.len(cur_namespace)+1)  --cut the namespace
+				if self._ele[rel_fieldname] then -- direct top-level assignment
+					self._ele[rel_fieldname].data.value = value
+				else
+					for elename, eledef in pairs(self._ele) do
+						if eledef.getContainerState then -- element supports sub-states
+							eledef:getContainerState():_sfs_process_value_(field, value)
+						end
+					end
 				end
 			end
-			for key,val in pairs(self._ele) do
-				if val.submit then
-					val:submit(fields, player)
+		end,
+		-- process action for received field if supported
+		_sfs_process_action_ = function(self, field, value, player)
+			local cur_namespace = self:getNamespace()
+			if cur_namespace == "" or cur_namespace == string.sub(field, 1, string.len(cur_namespace)) then -- Check current namespace
+				local rel_fieldname = string.sub(field, string.len(cur_namespace)+1) --cut the namespace
+				if self._ele[rel_fieldname] then -- direct top-level assignment
+					if self._ele[rel_fieldname].submit then
+						self._ele[rel_fieldname]:submit(value, player)
+					end
+				else
+					for elename, eledef in pairs(self._ele) do
+						if eledef.getContainerState then -- element supports sub-states
+						eledef:getContainerState():_sfs_process_action_(field, value, player)
+						end
+					end
 				end
 			end
-
-			-- call onInput hook if enabled
+		end,
+		-- process onInput hook for the state
+		_sfs_process_oninput_ = function(self, fields, player)
 			if self._onInput then
 				self:_onInput(fields, player)
 			end
-
+			-- recursive all onInput hooks on visible containers
```

Comment:
if you called on_receive_fields for each container then `if self._onInput` could be called from the top of on_receive_fields